Retrieve content (file, dir, symlink)

Palleas opened the pull request:
I know that we shouldn't spend too much time adding new features now that the GraphQL API is out there **but** I need to be able to fetch a file from a repository so... here we are 😄 

I only have one question left, should we handle decoding the `base64` content? 🤔

ikesyo commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,84 @@
+    public let url: URL
+
+    /// URL to download the content when content is a file or content is a symlink to a file
+    public let downloadUrl: URL?
```
This should be `downloadURL`: https://swift.org/documentation/api-design-guidelines/#follow-case-conventions.

ikesyo commented on Tentacle/Client.swift:
```diff
@@ -312,6 +317,11 @@ public final class Client {
     }
 
+    /// Fetch a content in a specific repository
+    public func content(in repository: Repository, at path: String) -> SignalProducer<(Response, Content), Error> {
```
> `at path: String`

This should be `atPath path: String` for the follwing reasons:

- https://swift.org/documentation/api-design-guidelines/#weak-type-information
- https://developer.apple.com/reference/foundation/filemanager

mdiep replied:
It might make more sense to put the path first. I also think it could just be called `path`.

```
content(path: String, in repository: Repository)
```

mdiep replied:
This doesn't seem to work if the path refers to a directory.

ikesyo commented on Tentacle/Client.swift:
```diff
@@ -191,6 +194,8 @@ public final class Client {
                 return "/orgs/\(organisation)/repos"
             case .publicRepositories:
                 return "/repositories"
+            case .content(let owner, let repository, let path):
```
Could you please write this as follows for consistency?

```swift
case let .content(owner, repository, path):
```

ikesyo commented on Tentacle/Client.swift:
```diff
@@ -463,6 +473,8 @@ extension Client.Endpoint: Hashable {
             return organisation.hashValue
         case .publicRepositories:
             return "PublicRepositories".hashValue
+        case .content(let owner, let repository, let path):
```
Ditto

mdiep commented on Tentacle/Client.swift:
```diff
@@ -312,6 +317,11 @@ public final class Client {
         return fetchMany(.publicRepositories, page: page, pageSize: perPage)
     }
 
+    /// Fetch a content in a specific repository
```
I'd write:

> Fetch the content for a path in the repository

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,84 @@
+    public enum ContentType: String {
+        case file
+        case dir
+        case symlink
```
It seems that there should be a `submodule` case here.

mdiep commented on Tests/ContentTests.swift:
```diff
@@ -0,0 +1,29 @@
+        )
+
+        XCTAssertEqual(Fixture.FileForRepository.ReadMeForSampleRepository.decode()!, expected)
+    }
```
I think we should also have tests for a directory, a symlink, and a submodule.

Palleas commented:
The API gives you a JSON object when you're targeting a file and an array when it's a directory... Any idea on how to handle that? I was thinking 2 methods: instead of one `content`, a `file` and a `files`..? Not sure how clean it feels.

mdiep commented:
Otherwise we could use an `Either` type or make `Content` an enum.

```swift
enum Content {
    struct File {
        ...
    }
    case file(File)
    case directory([Content])
}
```

Palleas commented:
Now supporting submodules, symlinks, directories... Let me know what you think :shipit:

Palleas commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,134 @@
+
+extension Content.File: Hashable {
+    public static func ==(lhs: Content.File, rhs: Content.File) -> Bool {
+        return lhs.name == rhs.name
```
I need a more complex `==` here, at least based on `path` and `sha`

Palleas commented:
Rebased on master and better `==`, ready for review!

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,136 @@
+        public let type: ContentType
+
+        /// Path to actual content when content is a symlink
+        public let target: String?
```
It seems like we should move this into the innermost `enum`.

```swift
struct File {
    enum Content {
        case file(size: Int)
        case directory
        case symlink(target: String)
        case submodule(url: String)
    }

    let content: Content
}
```

What do you think?

Palleas replied:
That makes more sense!

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,136 @@
+            case file
+            case dir
+            case symlink
+            case submodule
```
I think we should make the URL of the submodule visible if possible. Also, we should have a test for a directory that contains a submodule because of [this behavior](https://developer.github.com/v3/repos/contents/#response-if-content-is-a-directory). I'm curious if we can detect that case by looking for `submodule_git_url` or something (instead of switching on the actual type that the API returns).

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,136 @@
+                return .failure(.custom("\(error)"))
+            }
+        case .object(_):
+            switch Content.File.decode(j) {
```
Can you use `map` on the `Decode` here and above? `return Content.File.decode(j).map(Content.file)`

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,136 @@
+    public static func ==(lhs: Content.File, rhs: Content.File) -> Bool {
+        return lhs.name == rhs.name
+            && lhs.path == rhs.path
+            && lhs.sha == rhs.sha
```
This should probably take the type into account as well. The `sha` should do this 99.99999% of the time, but it never hurts to be cautious.

mdiep commented on Tests/ContentTests.swift:
```diff
@@ -0,0 +1,86 @@
+                sha: "e3fe7edcb2247a69c6eade1719d1e0cd83595cf9",
+                url: URL(string: "https://github.com/mdiep/Tentacle/blob/master/update-test-fixtures/main.swift")!,
+                downloadURL: URL(string: "https://raw.githubusercontent.com/mdiep/Tentacle/master/update-test-fixtures/main.swift")
+            )
```
I think it'd be good for this test to have one of each type of file.

Palleas commented:
I'll push an update later today! 👍

Palleas commented:
Just to be clear, this is what you had in mind? 
```swift
public struct File {
    public enum Content {
        case file(size: Int)
        case directory([File])
        case symlink(target: String)
        case submodule(url: String)
    }

    /// Name of the file
    public let name: String

    /// Path to the file in repository
    public let path: String

    /// Sha of the file
    public let sha: String

    /// URL to preview the content
    public let url: URL

    /// URL to download the content when content is a file or content is a symlink to a file
    public let downloadURL: URL?

    /// The type of content
    public let content: Content

    public var description: String {
        return name
    }
}
```

mdiep commented:
I think it needs to be:

```swift
public enum Content {
    public struct File: CustomStringConvertible {
        public enum Content {
            case file(size: Int)
            case directory
            case symlink(target: String)
            case submodule(url: String)
        }

        /// Name of the file
        public let name: String

        /// Path to the file in repository
        public let path: String

        /// Sha of the file
        public let sha: String

        /// URL to preview the content
        public let url: URL

        /// URL to download the content when content is a file or content is a symlink to a file
        public let downloadURL: URL?

        /// The type of content
        public let content: Content

        public var description: String {
            return name
        }
    }
    
    case file(File)
    case directory([File])
}
```

The outermost `directory` is for the return value. The innermost one is if the directory contains a directory.

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,194 @@
+        public let url: URL
+
+        /// URL to download the content when content is a file or content is a symlink to a file
+        public let downloadURL: URL?
```
Should we move this into `Content.File.ContentType` as well?

mdiep commented:
I'm not sure if you were done, but I reviewed anyway. ☺️ There are also a couple unaddressed items above.

Palleas commented:
I've updated my PR so the directory we decode contains a file, a symlink and a submodule. Fun facts:
* a submodule in a directory seems to appear as a file
* a symlink in a directory doesn't have a target 

:thinking:

mdiep commented:
> a submodule in a directory seems to appear as a file

Yeah, that was mentioned in the documentation. That was why I wanted a test for a directory with each the of file. It looks like we could detect submodules though: they have a `null` `download_url`. What do you think of detecting based on that?

mdiep commented on Tests/ContentTests.swift:
```diff
@@ -0,0 +1,78 @@
+
+    func testDecodedDirectory() {
+        let expected: Content = .directory([
+            Content.File(
```
We should probably have a directory here in too? There might be additional gotchas.

Palleas commented:
> What do you think of detecting based on that?

That could work but I don't have the submodule of the git repository thought... Is that ok?

mdiep commented:
That seems better than reporting it as a file? 😕

Palleas commented:
It seems better now, I've updated the way submodules in directories are handled and I've added a directory in the Sample repository. I think we're hopefully maybe probably getting close to something usable..?

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,163 @@
+            case file(size: Int, downloadURL: URL?)
+            case directory
+            case symlink(target: String?, downloadURL: URL?)
+            case submodule(url: String?)
```
Can you document why these fields are optional? (i.e. that retrieving a directory doesn't get all the fields for all the files)

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,163 @@
+        case let (.file(size, url), .file(size2, url2)) where size == size2 && url == url2: return true
+        case (.directory, .directory): return true
+        case let (.submodule(url), .submodule(url2)) where url == url2: return true
+        case let (.symlink(target, url), .symlink(target2, url2)) where target == target2 && url == url2: return true
```
I'd prefer to write these like so:

```
case let (.symlink(target, url), .symlink(target2, url2)):
    return target == target2 && url == url2
```

mdiep commented on Tentacle/Content.swift:
```diff
@@ -0,0 +1,163 @@
+
+public enum Content {
+    public struct File: CustomStringConvertible {
+        public enum ContentType {
```
These types could all use some documentation.

mdiep commented:
🎉

Palleas commented:
:tada: Thanks! It took a while but I learned a ton. ✌️
